internal/util/function/embedding/zilliz_embedding_provider.go

```diff
@@ -0,0 +1,107 @@
+/*
+ * # Licensed to the LF AI & Data foundation under one
+ * # or more contributor license agreements. See the NOTICE file
+ * # distributed with this work for additional information
+ * # regarding copyright ownership. The ASF licenses this file
+ * # to you under the Apache License, Version 2.0 (the
+ * # "License"); you may not use this file except in compliance
+ * # with the License. You may obtain a copy of the License at
+ * #
+ * #     http://www.apache.org/licenses/LICENSE-2.0
+ * #
+ * # Unless required by applicable law or agreed to in writing, software
+ * # distributed under the License is distributed on an "AS IS" BASIS,
+ * # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * # See the License for the specific language governing permissions and
+ * # limitations under the License.
+ */
+
+package embedding
+
+import (
+	"context"
+	"strings"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus/internal/util/function/models"
+	"github.com/milvus-io/milvus/internal/util/function/models/zilliz"
+	"github.com/milvus-io/milvus/pkg/v2/util/typeutil"
+)
+
+type ZillizEmbeddingProvider struct {
+	fieldDim int64
+
+	client    *zilliz.ZillizClient
+	modelName string
+
+	maxBatch    int
+	timeoutSec  int64
+	modelParams map[string]string
+	extralInfo  map[string]string
+}
+
+func NewZillizEmbeddingProvider(fieldSchema *schemapb.FieldSchema, functionSchema *schemapb.FunctionSchema, params map[string]string, extralInfo map[string]string) (*ZillizEmbeddingProvider, error) {
+	fieldDim, err := typeutil.GetDim(fieldSchema)
+	if err != nil {
+		return nil, err
+	}
+
+	var modelDeploymentID string
+	modelParams := map[string]string{}
+	for _, param := range functionSchema.Params {
+		switch strings.ToLower(param.Key) {
+		case models.ModelDeploymentIDKey:
+			modelDeploymentID = param.Value
+		default:
+			modelParams[param.Key] = param.Value
+		}
+	}
+
+	cluster := extralInfo[models.ClusterIDKey]
+	c, err := zilliz.NewZilliClient(modelDeploymentID, cluster, params)
+	if err != nil {
+		return nil, err
+	}
+
+	maxBatch := 64
+	provider := ZillizEmbeddingProvider{
+		client:      c,
+		fieldDim:    fieldDim,
+		maxBatch:    maxBatch,
+		timeoutSec:  30,
+		modelParams: modelParams,
+		extralInfo:  extralInfo,
+	}
+	return &provider, nil
+}
+
+func (provider *ZillizEmbeddingProvider) MaxBatch() int {
+	return 5 * provider.maxBatch
+}
+
+func (provider *ZillizEmbeddingProvider) FieldDim() int64 {
+	return provider.fieldDim
+}
+
+func (provider *ZillizEmbeddingProvider) CallEmbedding(ctx context.Context, texts []string, mode models.TextEmbeddingMode) (any, error) {
+	numRows := len(texts)
+	if mode == models.SearchMode {
+		provider.modelParams["input_type"] = "query"
```

Comment:
If a provider get called concurrently, this will be a race. But I guess it will not right?